### ISSUE TITLE
Inconsistent scale selection behavior for the same `args.multi_scale and not args.do_random_resize_via_padding` condition.

### ISSUE THREAD

**Issue by Abdul-Mukit** [roboflow/rf-detr#327]
During training, the scale selection is dependent on the logic `args.multi_scale and not args.do_random_resize_via_padding`. However, the decision about scale selection is inconsistent between coco.py and engine.py. 

In coco.py `make_coco_transforms_square_div_64`, we have the following:
```
    scales = [resolution]
    if multi_scale:
        # scales = [448, 512, 576, 640, 704, 768, 832, 896]
        scales = compute_multi_scale_scales(resolution, expanded_scales, patch_size, num_windows)
        if skip_random_resize:
            scales = [scales[-1]]
        print(scales)

    if image_set == 'train':
        return T.Compose([
            T.RandomHorizontalFlip(),
            T.RandomSelect(
                T.SquareResize(scales),
                T.Compose([
                    T.RandomResize([400, 500, 600]),
                    T.RandomSizeCrop(384, 600),
                    T.SquareResize(scales),
                ]),
            ),
            normalize,
        ])
```
The argument, skip_random_resize = not args.do_random_resize_via_padding.

Here is how we call `make_coco_transforms_square_div_64`, in `build_roboflow`.
```
        dataset = CocoDetection(img_folder, ann_file, transforms=make_coco_transforms_square_div_64(
            image_set,
            resolution,
            multi_scale=args.multi_scale,
            expanded_scales=args.expanded_scales,
            skip_random_resize=not args.do_random_resize_via_padding,
            patch_size=args.patch_size,
            num_windows=args.num_windows
        ))
```

Due to `(args.multi_scale and not args.do_random_resize_via_padding) == True`, we select only one scale in `make_coco_transforms_square_div_64`. I am experimenting with the medium model, so scale becomes [736].
The argument already says multi_scale = True. But we don't respect that in make_coco_transforms_square_div_64.

Now, in enginer.py's `train_one_epoch` we have:
```
        if args.multi_scale and not args.do_random_resize_via_padding:
            scales = compute_multi_scale_scales(args.resolution, args.expanded_scales, args.patch_size, args.num_windows)
            random.seed(it)
            scale = random.choice(scales)
            with torch.inference_mode():
                samples.tensors = F.interpolate(samples.tensors, size=scale, mode='bilinear', align_corners=False)
                samples.mask = F.interpolate(samples.mask.unsqueeze(1).float(), size=scale, mode='nearest').squeeze(1).bool()

```
Note that although the exact same condition has been met, `(args.multi_scale and not args.do_random_resize_via_padding)`, this time we decide to respect args.multi_scale and populate scales with several different scale-values to randomly choose from. Eventually, we do perform this random rescaling. 

This creates a problem. Now, for example, when loading mask-annotations, we have to manually implement every transform in this block during the train_one_epoch. This is unnecessary and can be avoided easily by commenting out the ` if skip_random_resize:` section in `make_coco_transforms_square_div_64`. 
We will need to modify the `collate_fn` to something like this to make sure we cover for masks as well.
```
def collate_fn(batch):
    batch = list(zip(*batch))
    batch[0] = nested_tensor_from_tensor_list(batch[0])

    if len(batch) == 0:
        return tuple(batch)

    # targets are available.
    padded_image_size = batch[0].tensors.shape[-2:]
    for idx, target in enumerate(batch[1]):
        if "masks" not in target:
            continue

        unpadded_mask_size = target['masks'].shape[-2:]
        if padded_image_size == unpadded_mask_size:
            continue

        padding = [
            padded_image_size[0] - unpadded_mask_size[0],
            padded_image_size[1] - unpadded_mask_size[1]
        ]

        target['masks'] = torch.nn.functional.pad(
            target['masks'],
            (0, padding[0], 0, padding[1])
        )

    return tuple(batch)
```

My point is:
We should fix coco.py to respect args.multi_scale. The current fix in engine.py's `train_one_epoch` feels like a hack to cover for the initial mistake in coco.py.

**Comment by Abdul-Mukit** [comment]
Yes, I am trying to experiment with instance segmentations.

**Comment by Abdul-Mukit** [comment]
> which leads to slower training and worse results
Got it, thanks.
